pkg/writers/buffered_file_writer/bufferedfilewriter.go

```diff
@@ -13,12 +13,65 @@ import (
 	"github.com/trufflesecurity/trufflehog/v3/pkg/context"
 )
 
-// bufferPool is used to store buffers for reuse.
-var bufferPool = sync.Pool{
-	// TODO: Consider growing the buffer before returning it if we can find an optimal size.
-	// Ideally the size would cover the majority of cases without being too large.
-	// This would avoid the need to grow the buffer when writing to it, reducing allocations.
-	New: func() any { return new(bytes.Buffer) },
+type bufPoolOpt func(pool *bufferPool)
+
+type bufferPool struct {
+	bufferSize uint32
+	*sync.Pool
+}
+
+const defaultBufferSize = 2 << 12 // 8KB
+func newBufferPool(opts ...bufPoolOpt) *bufferPool {
+	pool := &bufferPool{bufferSize: defaultBufferSize}
+
+	for _, opt := range opts {
+		opt(pool)
+	}
+	pool.Pool = &sync.Pool{
+		New: func() any {
+			buf := new(bytes.Buffer)
+			buf.Grow(int(pool.bufferSize))
+			return buf
+		},
+	}
+
+	return pool
+}
+
+// sharedBufferPool is the shared buffer pool used by all BufferedFileWriters.
+// This allows for efficient reuse of buffers across multiple writers.
+var sharedBufferPool *bufferPool
+
+func init() { sharedBufferPool = newBufferPool() }
+
+func (bp *bufferPool) get(ctx context.Context) *bytes.Buffer {
+	buf, ok := bp.Pool.Get().(*bytes.Buffer)
+	if !ok {
+		ctx.Logger().Error(fmt.Errorf("buffer pool returned unexpected type"), "using new buffer")
+		buf = bytes.NewBuffer(make([]byte, 0, bp.bufferSize))
+	}
+
+	return buf
+}
+
+func (bp *bufferPool) growBufferWithSize(buf *bytes.Buffer, size int) {
+	// Grow the buffer to accommodate the new data.
+	buf.Grow(size)
+}
+
+func (bp *bufferPool) put(buf *bytes.Buffer) {
+	// If the buffer is more than twice the default size, replace it with a new, smaller one.
+	// This prevents us from returning very large buffers to the pool.
+	const maxAllowedCapacity = 2 * defaultBufferSize
+	if buf.Cap() > maxAllowedCapacity {
+		// Replace the buffer with a new, smaller one. No need to copy data since we're resetting it.
+		buf = bytes.NewBuffer(make([]byte, 0, defaultBufferSize))
```

Comment:
Does this mean large buffers get garbage collected?

Reply:
We could also explicitly set the buffer to nil.. maybe i'll do that so it's obvious.